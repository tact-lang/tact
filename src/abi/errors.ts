// Errors 0-127 are reserved for VM errors
// Errors 128-255 are reserved for contract errors
export const contractErrors = {
    null: { id: 128, message: "Null reference exception" },
    invalidPrefix: { id: 129, message: "Invalid serialization prefix" },
    invalidMessage: { id: 130, message: "Invalid incoming message" },
    constraintsError: { id: 131, message: "Constraints error" },
    accessDenied: { id: 132, message: "Access denied" },
    contractStopped: { id: 133, message: "Contract stopped" },
    invalidArgument: { id: 134, message: "Invalid argument" },
    codeNotFound: { id: 135, message: "Code of a contract was not found" },
    invalidAddress: { id: 136, message: "Invalid address" },
    masterchainNotEnabled: {
        id: 137,
        message: "Masterchain support is not enabled for this contract",
    },
};
