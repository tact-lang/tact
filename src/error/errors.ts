import path from "path";
import { cwd } from "process";
import { AstFuncId, AstId, AstTypeId } from "../ast/ast";
import { SrcInfo } from "../grammar";

export class TactError extends Error {
    readonly loc?: SrcInfo;
    constructor(message: string, loc?: SrcInfo) {
        super(message);
        this.loc = loc;
    }
}

// Any regular compilation error shown to user:
// parsing, typechecking, code generation
export class TactCompilationError extends TactError {
    constructor(message: string, loc?: SrcInfo) {
        super(message, loc);
    }
}

export class TactInternalCompilerError extends TactError {
    constructor(message: string, loc?: SrcInfo) {
        super(message, loc);
    }
}

export class TactConstEvalError extends TactCompilationError {
    fatal: boolean = false;
    constructor(message: string, fatal: boolean, loc: SrcInfo) {
        super(message, loc);
        this.fatal = fatal;
    }
}

export function locationStr(sourceInfo: SrcInfo): string {
    if (sourceInfo.file) {
        const loc = sourceInfo.interval.getLineAndColumn() as {
            lineNum: number;
            colNum: number;
        };
        const file = path.relative(cwd(), sourceInfo.file);
        return `${file}:${loc.lineNum}:${loc.colNum}: `;
    } else {
        return "";
    }
}

export function throwCompilationError(
    message: string,
    source?: SrcInfo,
): never {
    const msg =
        source === undefined
            ? message
            : `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`;
    throw new TactCompilationError(msg, source);
}

export function throwInternalCompilerError(
    message: string,
    source?: SrcInfo,
): never {
    const msg = `[INTERNAL COMPILER ERROR]: ${message}\nPlease report at https://github.com/tact-lang/tact/issues`;
    throw source === undefined
        ? new TactInternalCompilerError(msg)
        : new TactInternalCompilerError(
              `${locationStr(source)}\n${msg}\n${source.interval.getLineAndColumnMessage()}`,
              source,
          );
}

export function throwConstEvalError(
    message: string,
    fatal: boolean,
    source: SrcInfo,
): never {
    throw new TactConstEvalError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        fatal,
        source,
    );
}

export function idTextErr(
    ident: AstId | AstFuncId | AstTypeId | string,
): string {
    if (typeof ident === "string") {
        return `"${ident}"`;
    }
    return `"${ident.text}"`;
}

export type TactErrorCollection =
    | Error
    | TactCompilationError
    | TactInternalCompilerError
    | TactConstEvalError;
