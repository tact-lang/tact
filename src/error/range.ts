/**
 * Range of source code positions
 * 0-based
 * `end` points to position just after the end of range
 */
export type Range = {
    start: number;
    end: number;
};

/**
 * Intersect ranges
 */
export const intersect = (a: Range, b: Range): Range => {
    return {
        start: Math.max(a.start, b.start),
        end: Math.min(a.end, b.end),
    };
};

/**
 * Shift both sides of range by scalar
 */
export const shift = (a: Range, b: number) => {
    return {
        start: a.start + b,
        end: a.end + b,
    };
};
