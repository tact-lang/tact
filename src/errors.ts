import { MatchResult } from "ohm-js";
import path from "path";
import { cwd } from "process";
import { ASTRef } from "./grammar/ast";

export class TactError extends Error {
    readonly ref: ASTRef;
    constructor(message: string, ref: ASTRef) {
        super(message);
        this.ref = ref;
    }
}

export class TactParseError extends TactError {
    constructor(message: string, ref: ASTRef) {
        super(message, ref);
    }
}

/// This will be split at least into two categories: typechecking and codegen errors
export class TactCompilationError extends TactError {
    constructor(message: string, ref: ASTRef) {
        super(message, ref);
    }
}

export class TactConstEvalError extends TactCompilationError {
    fatal: boolean = false;
    constructor(message: string, fatal: boolean, ref: ASTRef) {
        super(message, ref);
        this.fatal = fatal;
    }
}

function locationStr(sourceInfo: ASTRef): string {
    if (sourceInfo.file) {
        const loc = sourceInfo.interval.getLineAndColumn() as {
            lineNum: number;
            colNum: number;
        };
        const file = path.relative(cwd(), sourceInfo.file);
        return `${file}:${loc.lineNum}:${loc.colNum}: `;
    } else {
        return "";
    }
}

export function throwParseError(matchResult: MatchResult, path: string): never {
    const interval = matchResult.getInterval();
    const source = new ASTRef(interval, path);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const message = `Parse error: expected ${(matchResult as any).getExpectedText()}\n`;
    throw new TactParseError(
        `${locationStr(source)}${message}\n${interval.getLineAndColumnMessage()}`,
        source,
    );
}

export function throwCompilationError(message: string, source: ASTRef): never {
    throw new TactCompilationError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        source,
    );
}

export function throwConstEvalError(
    message: string,
    fatal: boolean,
    source: ASTRef,
): never {
    throw new TactConstEvalError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        fatal,
        source,
    );
}
