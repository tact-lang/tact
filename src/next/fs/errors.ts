import type { Logger } from "@/error/logger-util";

export const FsErrors = <M, R>(l: Logger<M, R>) => {
    const regular = {
        ENOENT: (absPath: string) => {
            return l.error(l.text`File not found: ${l.path(absPath)}`);
        },
        EACCES: (absPath: string) => {
            return l.error(l.text`Permission denied: ${l.path(absPath)}`);
        },
        EISDIR: (absPath: string) => {
            return l.error(l.text`Directory is not a file: ${l.path(absPath)}`);
        },
        EMFILE: (absPath: string) => {
            return l.error(l.text`Too many open files: ${l.path(absPath)}`);
        },
        ENAMETOOLONG: (absPath: string) => {
            return l.error(l.text`Filename too long: ${l.path(absPath)}`);
        },
        ENOSPC: (absPath: string) => {
            return l.error(l.text`No space left on device: ${l.path(absPath)}`);
        },
    } as const;
    return {
        regular,
        unexpected: (error: unknown) => {
            throw error;
        },
        outOfRoot: (absPath: string) => {
            return l.error(
                l.text`Outside of the root directory: ${l.path(absPath)}`,
            );
        },
        readonly: (absPath: string) => {
            return l.error(l.text`Filesystem is readonly: ${l.path(absPath)}`);
        },
    };
};

export type FsErrors = ReturnType<typeof FsErrors>;
