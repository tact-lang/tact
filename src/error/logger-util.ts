import { catchUncolored } from "./async-util";
import { TactInternalError } from "./errors";
import type { Range } from "./range";

/**
 * Methods for formatting an error message
 */
export interface Formatter<M> {
    /**
     * Substitutions into template literal
     * Creates a single message, possible with some parts of it rendered in a
     * different way.
     */
    readonly text: (
        parts: TemplateStringsArray,
        ...subst: readonly (string | M)[]
    ) => M;

    /**
     * Absolute path token
     * Displayed in blue in CLI and as link in IDE
     * Displayed relative to `cwd` in CLI or workspace root in IDE
     * External tooling can use absolute path in other ways
     */
    readonly path: (path: string) => M;

    /**
     * Refer to `id` in a `path` file at `range`
     * In IDE should display as a link with `id` as text
     */
    readonly locatedId: (id: string, path: string, range: Range) => M;

    /**
     * Token for a list of expected strings
     */
    readonly expected: (guesses: ReadonlySet<string>) => M;
}

/**
 * Basic logger interface
 */
export interface BaseLogger<M, R> {
    /**
     * Throw internal error
     *
     * If internal error is not related to currently source and range,
     * `internal()` from `errors.ts` will work as well
     */
    readonly internal: (message: M) => never;

    /**
     * Throw compilation error
     */
    readonly error: (message: M) => R;

    /**
     * Display a warning
     */
    readonly warn: (message: M) => void;

    /**
     * Log something not that important
     */
    readonly info: (message: M) => void;

    /**
     * Even if using error recovery, exit compilation right now and there
     *
     * Used at the end of compilation stage, if continuing compilation is
     * considered dangerous due to potentially broken internal compiler state
     */
    readonly exitIfErrored: () => void;
}

/**
 * (Co-)map message types of base logger
 *
 * Used when we need to add something to every message
 */
const mapBaseLogger = <M1, M2, R>(
    log: BaseLogger<M1, R>,
    f: (m: M2) => M1,
): BaseLogger<M2, R> => ({
    exitIfErrored: log.exitIfErrored,
    internal: (m) => log.internal(f(m)),
    error: (m) => log.error(f(m)),
    warn: (m) => {
        log.warn(f(m));
    },
    info: (m) => {
        log.info(f(m));
    },
});

/**
 * Logger that knows about currently compiled file
 */
export interface SourceLogger<M, R> extends Formatter<M>, BaseLogger<M, R> {
    /**
     * Choose range where an error will be shown
     */
    at: (range: Range) => BaseLogger<M, R>;

    /**
     * Handle multiple errors in callback without throwing them
     */
    recover: <T>(cb: (logger: SourceLogger<M, void>) => T) => T;
}

/**
 * Top-level logger
 */
export interface Logger<M, R> extends Formatter<M>, BaseLogger<M, R> {
    /**
     * Set currently compiled source in logging context
     */
    source: <T>(
        path: string,
        code: string,
        cb: (logger: SourceLogger<M, R>) => T,
    ) => T;

    /**
     * Handle multiple errors in callback without throwing them
     */
    recover: <T>(cb: (logger: Logger<M, void>) => T) => T;
}

/**
 * Error used to stop compilation
 *
 * @private Do not use outside of this file!
 */
export class _ExitError extends Error {}

/**
 * Stop compilation
 *
 * @private Do not use outside of this file!
 */
export const _exit = () => {
    throw new _ExitError();
};

/**
 * Do not stop compilation
 *
 * @private Do not use outside of this file!
 */
export const _ignore = () => {};

/**
 * Used internally to handle errors that bubbled to root logger
 */
export const handleTopLevelErrors = <M, T, U>(
    log: Logger<M, unknown>,
    cb: () => T,
    onExit: () => U,
) => {
    // `catchUncolored` is instrumental here, to allow `T` to be either
    // regular value or a `Promise`
    return catchUncolored<T | U>(
        () =>
            catchUncolored(cb, (e) => {
                if (e instanceof _ExitError) {
                    // Exit will be handled below
                    throw e;
                } else if (e instanceof TactInternalError) {
                    // Format `throwInternal()` and rethrow it as regular error
                    // Will throw ExitError
                    return log.internal(log.text`${e.formattedMessage}`);
                } else {
                    // For any other error, convert it to string and rethrow as
                    // regular error. Will throw ExitError
                    const text = e instanceof Error ? e.toString() : String(e);
                    return log.internal(log.text`Unhandled: ${text}`);
                }
            }),
        (e) => {
            if (e instanceof _ExitError) {
                return onExit();
            } else {
                // impossible, but throw anyway
                throw e;
            }
        },
    );
};

/**
 * Messages of errors that are `throw`n should at least be expanded with
 * context, in while file they happened
 */
export const rethrowWithPath = (error: unknown, path: string) => {
    if (error instanceof TactInternalError) {
        error.formattedMessage = `${error.formattedMessage}\nwhile compiling ${path}`;
    } else if (error instanceof Error) {
        error.message = `${error.message}\nwhile compiling ${path}`;
    }
    throw error;
};

/**
 * Loggers need both single-error and multi-error loggers
 * Depending on `onError` handler (`exit` or `ignore`), they will do either
 *
 * `createChildLoggers` creates another error logging scope, that
 * computes its own `hadErrors` flag, and also sets `hadErrors` on all of its
 * parent scopes.
 */
type BaseLoggers<M> = <R>(onError: () => R) => BaseLogger<M, R>;
const createChildLoggers = <M>(getBase: BaseLoggers<M>): BaseLoggers<M> => {
    let hadErrors = false;
    return (onError) => {
        const base = getBase(onError);
        return {
            ...base,
            exitIfErrored: () => {
                if (hadErrors) return _exit();
                // If this scope is not in error state, maybe its parent is
                base.exitIfErrored();
            },
            internal: (m) => {
                // We won't use this value, because `.internal()` immediately
                // exits the application anyway, but it's better to be consistent
                hadErrors = true;
                return base.internal(m);
            },
            error: (m) => {
                hadErrors = true;
                return base.error(m);
            },
        };
    };
};

/**
 * Create top-level single-error and multi-error loggers
 */
const makeBaseLoggers = <M>(iface: LoggerHandlers<M, unknown>) =>
    createChildLoggers<M>((onError) => ({
        exitIfErrored: () => {},
        internal: (message: M) => {
            iface.internal(message);
            // Internal errors directly exit the application
            return _exit();
        },
        error: (message: M) => {
            iface.error(message);
            // Behavior of this call distinguished single and multiple errors
            // if `onError` throws, it's a single error
            return onError();
        },
        warn: iface.warn,
        info: iface.info,
    }));

/**
 * Set of handlers that are different between logger implementations
 *
 * This interface is a simplified version of logger, so that we can implement
 * a new kind of logger without copying complex error message builders
 */
export interface LoggerHandlers<M, R> {
    readonly internal: (message: M) => void;
    readonly error: (message: M) => void;
    readonly warn: (message: M) => void;
    readonly info: (message: M) => void;
    readonly text: (
        parts: TemplateStringsArray,
        ...subst: readonly (string | M)[]
    ) => M;
    readonly path: (path: string) => M;
    readonly locatedId: (id: string, path: string, range: Range) => M;
    readonly expected: (guesses: ReadonlySet<string>) => M;
    readonly atPath: (path: string, message: M) => M;
    readonly atRange: (
        path: string,
        code: string,
        range: Range,
        message: M,
    ) => M;
    readonly onExit: () => R;
}

/**
 * Create user-facing logger from its internal representation and
 * application entrypoint
 */
export const makeLogger = <M, T>(
    iface: LoggerHandlers<M, T>,
    compile: (log: Logger<M, never>) => T,
) => {
    // Create a simple text logger
    const formatter: Formatter<M> = {
        text: iface.text,
        path: iface.path,
        expected: iface.expected,
        locatedId: iface.locatedId,
    };

    // Create a logger with a given source file in context
    const makeSourceLogger = <R>(
        path: string,
        code: string,
        baseLoggers: BaseLoggers<M>,
        handle: () => R,
    ): SourceLogger<M, R> => ({
        ...formatter,
        ...mapBaseLogger(baseLoggers(handle), (m) => iface.atPath(path, m)),
        at: (range) =>
            mapBaseLogger(baseLoggers(handle), (m) =>
                iface.atRange(path, code, range, m),
            ),
        recover: (cb) => cb(makeSourceLogger(path, code, baseLoggers, _ignore)),
    });

    // Create a function that returns a logger with source file context, and also
    // add context to all the native JS errors that pass through it
    const makeSourceFunction =
        <R>(baseLoggers: BaseLoggers<M>, handle: () => R) =>
        <T>(
            path: string,
            code: string,
            cb: (logger: SourceLogger<M, R>) => T,
        ): T => {
            // `catchUncolored` is instrumental here, to allow `T` to be either
            // regular value or a `Promise`
            return catchUncolored(
                () =>
                    cb(
                        makeSourceLogger(
                            path,
                            code,
                            createChildLoggers(baseLoggers),
                            handle,
                        ),
                    ),
                (error) => rethrowWithPath(error, path),
            );
        };

    // Create top-level logger, for one or several error messages, depending on `handle`
    const makeLogger = <R>(
        baseLoggers: BaseLoggers<M>,
        handle: () => R,
    ): Logger<M, R> => ({
        ...formatter,
        ...baseLoggers(handle),
        source: makeSourceFunction(baseLoggers, handle),
        recover: makeRecover(baseLoggers),
    });

    // Create a top-level multi-error logger
    const makeRecover =
        (baseLoggers: BaseLoggers<M>) =>
        <T>(cb: (logger: Logger<M, void>) => T): T => {
            return cb(makeLogger(createChildLoggers(baseLoggers), _ignore));
        };

    // Create top-level one-error logger
    const logger = makeLogger(makeBaseLoggers(iface), _exit);

    // Run application, catch all uncaught/internal errors and dispatch them as errors in log
    return handleTopLevelErrors(
        logger,
        () => {
            const result = compile(logger);
            logger.exitIfErrored();
            return result;
        },
        iface.onExit,
    );
};
