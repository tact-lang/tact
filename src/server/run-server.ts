import { thenUncolored } from "../error/async-util";
import { Logger, LoggerHandlers, makeLogger } from "../error/logger-util";
import { Range } from "../error/range";

/**
 * Error level
 *
 * For `internal` errors a popup with link to GitHub issues should be shown
 */
export type ErrorLevel = "internal" | "error" | "warn" | "info";

export type LogEntry = {
    readonly level: ErrorLevel;
    readonly message: Message;
};

export type Message =
    | MessageText
    | MessagePath
    | MessageLocatedId
    | MessageExpected
    | MessageSource;

/**
 * Text with substitutions
 *
 * Format matches template literals in JS. For N >= 0,
 * - `parts` will always have odd number of elements N+1
 * - `subst` will always have even number of elements N
 *
 * parts: ['A', 'C', 'E'], subst: ['B', 'D'] -> `ABCDE`
 *
 * If `Message` is converted to simple text, `showTemplate`
 * can further convert this back to string.
 */
export type MessageText = {
    readonly kind: "text";
    readonly parts: readonly string[];
    readonly subst: readonly (string | Message)[];
};

/**
 * Display a path
 */
export type MessagePath = {
    readonly kind: "path";
    /**
     * Absolute path
     *
     * In compiler we don't know where this path will be shown,
     * so we can't make it relative to anything
     */
    readonly path: string;
};

/**
 * Display a reference to identifier at some location in source
 *
 * Should display as a link with `id` as text, with click sending to
 * `range` of `path`.
 */
export type MessageLocatedId = {
    readonly kind: "locatedId";
    /**
     * Link text
     */
    readonly id: string;
    /**
     * Absolute path
     *
     * In compiler we don't know where this path will be shown,
     * so we can't make it relative to anything
     */
    readonly path: string;
    /**
     * Range where `id` is defined, used or otherwise mentioned
     */
    readonly range: Range;
};

/**
 * A set of expected strings
 *
 * If it's just shown, and not used for autofix purposes,
 * use `showExpectedText` from `string-util` to convert it to string
 */
export type MessageExpected = {
    readonly kind: "expected";
    readonly values: ReadonlySet<string>;
};

/**
 * Tells in which file an error happened
 */
export type MessageSource = {
    readonly kind: "source";
    readonly path: string;
    readonly message: Message;
    readonly range: Range | undefined;
};

/**
 * Logger for running Tact as a library
 */
export const runServer = <T>(compile: (log: Logger<Message, never>) => T) => {
    const log: LogEntry[] = [];
    const jsonIface: LoggerHandlers<Message, undefined> = {
        internal: (message) => log.push({ level: "internal", message }),
        error: (message) => log.push({ level: "error", message }),
        warn: (message) => log.push({ level: "warn", message }),
        info: (message) => log.push({ level: "info", message }),
        path: (path) => ({ kind: "path", path }),
        locatedId: (id, path, range) => ({
            kind: "locatedId",
            id,
            path,
            range,
        }),
        expected: (values) => ({ kind: "expected", values }),
        text: (parts, ...subst) => ({ kind: "text", parts: [...parts], subst }),
        atPath: (path, message) => ({
            kind: "source",
            path,
            message,
            range: undefined,
        }),
        atRange: (path, code, range, message) => ({
            kind: "source",
            path,
            message,
            range,
        }),
        onExit: () => undefined,
    };
    return thenUncolored(makeLogger(jsonIface, compile), () => log);
};
