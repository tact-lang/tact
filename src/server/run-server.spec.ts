import { throwInternal } from "../error/errors";
import { Logger, SourceLogger } from "../error/logger-util";
import { Range } from "../error/range";
import { runServer } from "./run-server";

describe("runServer", () => {
    test("only first error without error recovery", () => {
        const result = runServer((log) => {
            log.error(log.text`Error 1`);
            log.error(log.text`Error 2`);
        });
        expect(result).toMatchSnapshot();
    });

    test("all info logs are logged", () => {
        const result = runServer((log) => {
            log.info(log.text`Info 1`);
            log.info(log.text`Info 2`);
        });
        expect(result).toMatchSnapshot();
    });

    test("raw internal error", () => {
        const result = runServer((_log) => {
            throwInternal(`OMG`);
        });
        expect(result).toMatchSnapshot();
    });

    test("logger internal error", () => {
        const result = runServer((log) => {
            log.internal(log.text`OMG`);
        });
        expect(result).toMatchSnapshot();
    });

    test("raw internal error in source", () => {
        const result = runServer((log) => {
            log.source("/foo/bar", "Hello, world", () => {
                throwInternal(`OMG`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("logger internal error in source", () => {
        const result = runServer((log) => {
            log.source("/foo/bar", "Hello, world", (log) => {
                log.internal(log.text`OMG`);
                log.info(log.text`Impossible`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("internal error in source at range", () => {
        const result = runServer((log) => {
            log.source("/foo/bar", "Hello, world", (log) => {
                log.at({ start: 3, end: 5 }).internal(log.text`OMG`);
                log.info(log.text`Impossible`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("uncaught error", () => {
        const result = runServer(() => {
            throw new Error("Uncaught!");
        });
        expect(result).toMatchSnapshot();
    });

    test("uncaught error in source", () => {
        const result = runServer((log) => {
            return log.source("/foo/bar", "Hello, world", () => {
                throw new Error("hehe");
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("multiple errors", () => {
        const result = runServer((log) => {
            log.recover((log) => {
                log.error(log.text`foo`);
                log.error(log.text`bar`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("exit on error", () => {
        const result = runServer((log) => {
            log.recover((log) => {
                log.error(log.text`foo`);
                log.error(log.text`bar`);
                log.exitIfErrored();
            });
            log.error(log.text`impossible`);
        });
        expect(result).toMatchSnapshot();
    });

    test("multiple errors inside source", () => {
        const result = runServer((log) => {
            log.source("/foo/bar", "Hello, world", (log) => {
                log.recover((log) => {
                    log.at({ start: 3, end: 5 }).error(log.text`foo`);
                    log.error(log.text`bar`);
                    log.exitIfErrored();
                });
                log.error(log.text`impossible`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("source inside multiple errors", () => {
        const result = runServer((log) => {
            log.recover((log) => {
                log.source("/foo/bar", "Hello, world", (log) => {
                    log.at({ start: 3, end: 5 }).error(log.text`foo`);
                    log.error(log.text`bar`);
                    log.exitIfErrored();
                });
                log.error(log.text`impossible`);
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("typed errors", () => {
        const fooBarSchema = <M, R>(l: Logger<M, R>) => ({
            fooError: () => l.error(l.text`Foo!`),
            barError: () => l.error(l.text`Bar!`),
        });
        const result = runServer((log) => {
            log.recover((log) => {
                const l = fooBarSchema(log);
                l.fooError();
                l.barError();
            });
        });
        expect(result).toMatchSnapshot();
    });

    test("typed errors for source", () => {
        const fooBarSchemaSrc = <M, R>(l: SourceLogger<M, R>) => ({
            fooError: (at: Range) => l.at(at).error(l.text`Foo!`),
            barError: () => l.error(l.text`Bar!`),
        });
        const result = runServer((log) => {
            log.source("/foo/bar", "Hello, world", (log) => {
                log.recover((log) => {
                    const l = fooBarSchemaSrc(log);
                    l.fooError({ start: 3, end: 5 });
                    l.barError();
                });
            });
        });
        expect(result).toMatchSnapshot();
    });
});
