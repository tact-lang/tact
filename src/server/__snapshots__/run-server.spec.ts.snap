// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runServer all info logs are logged 1`] = `
[
  {
    "level": "info",
    "message": {
      "kind": "text",
      "parts": [
        "Info 1",
      ],
      "subst": [],
    },
  },
  {
    "level": "info",
    "message": {
      "kind": "text",
      "parts": [
        "Info 2",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer exit on error 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "foo",
      ],
      "subst": [],
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "bar",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer internal error in source at range 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "OMG",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": {
        "end": 5,
        "start": 3,
      },
    },
  },
]
`;

exports[`runServer logger internal error 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "text",
      "parts": [
        "OMG",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer logger internal error in source 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "OMG",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": undefined,
    },
  },
]
`;

exports[`runServer multiple errors 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "foo",
      ],
      "subst": [],
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "bar",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer multiple errors inside source 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "foo",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": {
        "end": 5,
        "start": 3,
      },
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "bar",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": undefined,
    },
  },
]
`;

exports[`runServer only first error without error recovery 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "Error 1",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer raw internal error 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "text",
      "parts": [
        "",
        "",
      ],
      "subst": [
        "OMG",
      ],
    },
  },
]
`;

exports[`runServer raw internal error in source 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "text",
      "parts": [
        "",
        "",
      ],
      "subst": [
        "OMG
while compiling /foo/bar",
      ],
    },
  },
]
`;

exports[`runServer source inside multiple errors 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "foo",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": {
        "end": 5,
        "start": 3,
      },
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "bar",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": undefined,
    },
  },
]
`;

exports[`runServer typed errors 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "Foo!",
      ],
      "subst": [],
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "text",
      "parts": [
        "Bar!",
      ],
      "subst": [],
    },
  },
]
`;

exports[`runServer typed errors for source 1`] = `
[
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "Foo!",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": {
        "end": 5,
        "start": 3,
      },
    },
  },
  {
    "level": "error",
    "message": {
      "kind": "source",
      "message": {
        "kind": "text",
        "parts": [
          "Bar!",
        ],
        "subst": [],
      },
      "path": "/foo/bar",
      "range": undefined,
    },
  },
]
`;

exports[`runServer uncaught error 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "text",
      "parts": [
        "Unhandled: ",
        "",
      ],
      "subst": [
        "Error: Uncaught!",
      ],
    },
  },
]
`;

exports[`runServer uncaught error in source 1`] = `
[
  {
    "level": "internal",
    "message": {
      "kind": "text",
      "parts": [
        "Unhandled: ",
        "",
      ],
      "subst": [
        "Error: hehe
while compiling /foo/bar",
      ],
    },
  },
]
`;
